### PR TITLE
ESQL: Fix layout when aggregating with aliases

### PR DESCRIPTION
**Pull request by alex-spies**
Fix https://github.com/elastic/elasticsearch/issues/117770
Fix https://github.com/elastic/elasticsearch/issues/117784

https://github.com/elastic/elasticsearch/pull/117699 made changes to how we plan aggregations which were supposed to only trigger when a query contained a `CATEGORIZE`, but accidentally changed a code path that seems to only be required for interoperability with pre-8.13 nodes. Because of this, we didn't notice failing tests until the periodic bwc tests ran.

The code this PR fixes addresses situations where `Aggregate` plan nodes contained _aliases inside the aggregates_. On `main` and `8.x`, this is effectively an illegal state: since https://github.com/elastic/elasticsearch/pull/104958, aliases in the aggregates become `Eval` nodes before and after the `Aggregate` node.

But here, on 8.x, we'll just fix this code path so that it behaves exactly as before https://github.com/elastic/elasticsearch/pull/117699.

If this passes the full-bwc test, I plan to forward-port this by removing the obsolete code path on `main`.

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java [hunk 1]
```diff
@@ -120,10 +120,14 @@ public final PhysicalOperation groupingPhysicalOperation(
...
                  */
+                // TODO: This is likely required only for pre-8.14 node compatibility; confirm and remove if possible.
+                // Since https://github.com/elastic/elasticsearch/pull/104958, it shouldn't be possible to have aliases in the aggregates
+                // which the groupings refer to. Except for `BY CATEGORIZE(field)`, which remains as alias in the grouping, all aliases
+                // should've become EVALs before or after the STATS.
                 for (NamedExpression agg : aggregates) {
                     if (agg instanceof Alias a) {
                         if (a.child() instanceof Attribute attr) {
-                            if (groupAttribute.id().equals(attr.id())) {
+                            if (sourceGroupAttribute.id().equals(attr.id())) {
```
[the remark this answers is not in the document]
Comment (alex-spies) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java:130]: This line I just updated for alignment with the rest. It's the change below that actually fixes things.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java [hunk 2]
```diff
@@ -133,8 +137,8 @@ public final PhysicalOperation groupingPhysicalOperation(
                             // is in the output form
                             // if the group points to an alias declared in the aggregate, use the alias child as source
                             else if (aggregatorMode.isOutputPartial()) {
-                                if (groupAttribute.semanticEquals(a.toAttribute())) {
-                                    groupAttribute = attr;
+                                if (sourceGroupAttribute.semanticEquals(a.toAttribute())) {
+                                    sourceGroupAttribute = attr;
                                     break;
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java:142]: The `groupAttribute` was mutated, but we use the `sourceGroupAttribute` to determine the right channel to pass to the `groupSpecs`. We need to mutate the `sourceGroupAttribute` instead.

**Review by luigidellaquila — APPROVED** [review]
LGTM, good catch @alex-spies 

I think this will be fixed as well https://github.com/elastic/elasticsearch/issues/117861

**Review by ivancea — APPROVED** [review]
Looks good, thanks for quickly finding the issue!

**Comment by alex-spies** [comment]
The `full-bwc` task struggles with the `8.15.6` cases due to upstream issues:
```
* What went wrong:
Could not determine the dependencies of task ':qa:ccs-rolling-upgrade-remote-cluster:v8.15.6#fullUpgraded'.
> Could not resolve all dependencies for configuration ':qa:ccs-rolling-upgrade-remote-cluster:es_distro_extracted_testclusters-qa-ccs-rolling-upgrade-remote-cluster-v8.15.6-remote-0-8.15.6'.
   > Could not find elasticsearch-distribution:elasticsearch:8.15.6.
     Searched in the following locations:
       - https://artifacts-no-kpi.elastic.co/downloads/elasticsearch/elasticsearch-8.15.6-linux-x86_64.tar.gz
     Required by:
         project :qa:ccs-rolling-upgrade-remote-cluster

```
I think it's better to have this fix in + un-skip the tests, esp. since the affected code path should only matter to 8.11/8.12; therefore, I'll try to re-run the CI without full bwc; it already passed the bwc checks for everything except 8.15.

**Pull request merged**